way-cooler-bg/src/main.rs

```diff
@@ -0,0 +1,285 @@
+    // TODO Get the output size from way cooler somehow...
+    //let resolution = output.get_resolution()
+    //    .expect("Couldn't get output resolution");
+    //let (width, height) = (resolution.w as i32, resolution.h as i32);
+    let width = 16; let height = 9;
+    let size = (width * height) as i32;
+
+    // Write in the color coding to the surface
+    for _ in 0..size {
+        unsafe {
+            tmp.write_u32::<NativeEndian>(transmute(color.0))
+                .expect("Could not write to file")
+        }
+    }
+    tmp.flush()
+        .expect("Could not flush buffer");
+
+    // Create the buffer that is mem-mapped to the temp file descriptor
+    let pool = shm.create_pool(tmp.as_raw_fd(), size);
+    let background_buffer = pool.create_buffer(0, width, height, width, WlShmFormat::Argb8888);
+    // Tell Way Cooler not to set put this in the tree, treat as background
+
+    // Attach the buffer to the surface
+    background_surface.attach(Some(&background_buffer), 0, 0);
+    Ok(background_buffer)
+}
+
+/// Given the data from an image, writes it to a special Wayland surface
+/// which is then rendered as a background for Way Cooler.
+fn generate_image_background(path: &str, background_surface: &mut WlSurface,
+                                 env: &WaylandEnv) -> BufferResult {
+    // TODO support more formats, split into separate function
+    let image = open(path)
+        .unwrap_or_else(|_| {
+            println!("Could not open \"{:?}\"", path);
+            panic!("Could not open image file");
+        });
+    let mut image = image.to_rgba();
+    let width = image.width();
+    let height = image.height();
+    let stride = width * 4;
+    let size = stride * height;
+    // TODO Split this into its own function
+    {
+        let pixels = image.enumerate_pixels_mut();
+        for (_x, _y, pixel) in pixels {
+            let alpha = pixel[3] as u32;
+            pixel[0] = weird_math(pixel[0], alpha);
+            pixel[1] = weird_math(pixel[1], alpha);
+            pixel[2] = weird_math(pixel[2], alpha);
+
+            let tmp = pixel[2];
+            pixel[2] = pixel[0];
+            pixel[0] = tmp;
+        }
+    }
+    let vec = image.into_vec();
+    let mut tmp = tempfile::NamedTempFile::new().expect("Unable to create a tempfile.");
+    tmp.set_len(size as u64).expect("Could not truncate length of file");
+    tmp.write_all(&*vec).unwrap();
+
+
+    // TODO I want this shit outta here
+    let shm = env.shm.as_ref().map(|o| &o.0).unwrap();
+
+    // Create the surface we are going to write into
+
+    let pool = shm.create_pool(tmp.as_raw_fd(), size as i32);
+    let background_buffer = pool.create_buffer(0, width as i32, height as i32, stride as i32, WlShmFormat::Argb8888);
+    // Tell Way Cooler not to put this in the tree, treat as background
+    // TODO Make this less hacky by actually giving way cooler access to this thing...
```

Comment:
Another comment that is in the wrong place. This was talking about how we tell way cooler that this is supposed to be the background surface. Right now we are being _super_ hacky and setting the class name for the view to be "Background" which Way Cooler treats differently.

The _correct_ way (and the way eg Sway does it) would be to use a well defined wayland protocol (see [how sway does it](https://github.com/SirCmpwn/sway/blob/master/protocols/desktop-shell.xml#L10)) where we pass the surface we want to Way Cooler (and then it would display it).

Using dbus would also be fine, and probably the way we ultimately go with. Either way, to make this secure, you are correct and we should have Lua spawn this program with the appropriate credentials. 

I don't knew enough about how Lua / HLua works to set that up in a better way than I already have (which is admittedly pretty bad/hacky). If you can think of a better way right now, without dbus, to spawn the background feel free to make changes. We can add dbus "privileges" for the background later